app/src/org/commcare/tasks/FormRecordToFileTask.java

```diff
@@ -0,0 +1,257 @@
+            for (int i = 0; i < ids.size(); ++i) {
+                records[i] = storage.read(ids.elementAt(i));
+                results[i] = FormUploadUtil.FAILURE;
+            }
+
+            publishProgress(Localization.get("bulk.form.start"));
+
+            for (int i = 0; i < records.length; ++i) {
+                FormRecord record = records[i];
+                try {
+                    //If it's unsent, go ahead and send it
+                    if (FormRecord.STATUS_UNSENT.equals(record.getStatus())) {
+                        File folder;
+                        try {
+                            folder = new File(record.getPath(c)).getCanonicalFile().getParentFile();
+                        } catch (IOException e) {
+                            Logger.log(AndroidLogger.TYPE_ERROR_WORKFLOW, "Bizarre. Exception just getting the file reference. Not removing." + getExceptionText(e));
+                            continue;
+                        }
+
+                        //Good!
+                        //Time to transfer forms to storage!
+                        try {
+                            results[i] = getFileInstanceFromStorage(folder, new SecretKeySpec(record.getAesKey(), "AES"));
+                        } catch (FileNotFoundException e) {
+                            if (CommCareApplication._().isStorageAvailable()) {
+                                //If storage is available generally, this is a bug in the app design
+                                Logger.log(AndroidLogger.TYPE_ERROR_DESIGN, "Removing form record because file was missing|" + getExceptionText(e));
+                            } else {
+                                //Otherwise, the SD card just got removed, and we need to bail anyway.
+                                CommCareApplication._().reportNotificationMessage(NotificationMessageFactory.message(ProcessIssues.StorageRemoved), true);
+                                break;
+                            }
+                            continue;
+                        }
+                        if (results[i].intValue() == FormUploadUtil.FAILURE) {
+                            publishProgress("Failure during zipping process");
+                            return null;
+                        }
+                    }
+                } catch (SessionUnavailableException sue) {
+                    this.cancel(false);
+                    return null;
+                } catch (Exception e) {
+                    //Just try to skip for now. Hopefully this doesn't wreck the model :/
+                    Logger.log(AndroidLogger.TYPE_ERROR_DESIGN, "Totally Unexpected Error during form submission" + getExceptionText(e));
+                }
+            }
+
+            long result = 0;
+            for (int i = 0; i < records.length; ++i) {
+                if (results[i] > result) {
+                    result = results[i];
+                }
+            }

```

Comment:
This loop should be lift into a named helper method that gives additional context to what is going on.